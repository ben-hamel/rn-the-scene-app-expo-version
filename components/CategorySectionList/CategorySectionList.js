import {
  View,
  Text,
  SectionList,
  FlatList,
  StyleSheet,
  TouchableOpacity,
  Image,
} from "react-native";
import React, { useState, useEffect } from "react";
import { collection, getDocs } from "firebase/firestore";
import { db } from "../../lib/firebase";

import { SafeAreaView } from "react-native-safe-area-context";
import { useNavigation, useTheme } from "@react-navigation/native";

//TODO Delete after refactoring
// const Item = ({ title }) => (
//   <View style={styles.item}>
//     <Text style={styles.title}>{title}</Text>
//   </View>
// );

const CategorySectionList = () => {
  const navigation = useNavigation();
  const [categories, setCategories] = useState([]);
  const [loading, setLoading] = useState(true);

  const { colors } = useTheme();

  const styles = styling(colors);
  useEffect(() => {
    const getData = async () => {
      /**
       * Grab all documents from the categories collection
       */
      const querySnapshot = await getDocs(collection(db, "categories"));

      /**
       * create an array of objects from the querySnapshot
       */
      let arr_Data = [];
      querySnapshot.forEach((doc) => {
        /**
         * store document in new variable
         */
        const document_Category = doc.data();
        /**
         * change categories array to data array so its compatible with the SectionList.
         */
        document_Category.data = document_Category.categories;
        delete document_Category.categories;

        arr_Data.push(document_Category);
        console.log("Title", document_Category.data);
      });

      /**
       * set the categories state to the array of objects
       */
      setCategories(arr_Data);
      setLoading(false);
      console.log("categories retrieved");
    };

    getData();
  }, []);

  if (loading) {
    return (
      <SafeAreaView>
        <View>
          <Text style={{ color: colors.text }}>Loading</Text>
        </View>
      </SafeAreaView>
    );
  }

  /**
   * List Item Component for the SectionList renderitem
   */
  //TODO move outside component
  const ListItem = ({ item }) => {
    return (
      <TouchableOpacity
        onPress={() => navigation.navigate("CategoryDetailScreen", item.title)}
      >
        <View style={styles.item}>
          <Image
            source={{
              uri: item.uri,
            }}
            style={styles.itemPhoto}
            resizeMode="cover"
          />
          <Text style={styles.itemText}>{item.title}</Text>
        </View>
      </TouchableOpacity>
    );
  };

  return (
    <View>
      <SectionList
        renderSectionHeader={({ section, index }) => (
          <>
            <Text style={[styles.sectionHeader, { color: colors.text }]}>
              {section.title}
            </Text>
            <FlatList
              horizontal
              data={section.data}
              renderItem={({ item }) => <ListItem item={item} />}
              showsHorizontalScrollIndicator={false}
              keyExtractor={(item, index) => item + index}
            />
          </>
        )}
        renderItem={() => {
          return null;
        }}
        sections={categories}
        keyExtractor={(item, index) => item + index}
      />
    </View>
  );
};

const styling = (theme) =>
  StyleSheet.create({
    container: {
      flex: 1,
      backgroundColor: "#121212",
    },
    sectionHeader: {
      fontWeight: "800",
      fontSize: 18,
      marginTop: 20,
      marginLeft: 10,
      marginBottom: 5,
    },
    item: {
      margin: 10,
    },
    itemPhoto: {
      width: 200,
      height: 200,
    },
    itemText: {
      color: theme.text,
      marginTop: 5,
    },
    padding: {
      padding: 10,
    },
  });

export default CategorySectionList;
