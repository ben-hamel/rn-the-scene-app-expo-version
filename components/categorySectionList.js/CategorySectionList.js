import { View, Text, SectionList, FlatList, Button, StyleSheet, StatusBar, TouchableOpacity, Image } from 'react-native'
import React, { useState, useEffect } from 'react'
import { arrayRemove, collection, getDocs } from 'firebase/firestore';
import { db } from './../../config/firebase';

import { SafeAreaView } from "react-native-safe-area-context";
import { useNavigation } from "@react-navigation/native";

const Item = ({ title }) => (
    <View style={styles.item}>
        <Text style={styles.title}>{title}</Text>
    </View>
);



const CategorySectionList = () => {
    const navigation = useNavigation();
    const [categories, setCategories] = useState([]);
    const [loading, setLoading] = useState(true);

    useEffect(() => {
        const getData = async () => {
            /**
             * Grab all documents from the categories collection
             */
            const querySnapshot = await getDocs(collection(db, "categories"));

            /**
             * create an array of objects from the querySnapshot
             */
            let arr_Data = [];
            querySnapshot.forEach((doc) => {
                /**
                 * store document in new variable
                */
                document = doc.data()
                /**
                 * change categories array to data array so its compatible with the SectionList.
                 */
                document.data = document.categories
                delete document.categories

                arr_Data.push(document);
                // console.log("TEST", datah);
            });

            /**
             * set the categories state to the array of objects
             */
            setCategories(arr_Data);
            setLoading(false);
        };

        getData();
    }, []);


    if (loading) {
        return (
            <SafeAreaView>
                <View>
                    <Text>Loading</Text>
                </View>
            </SafeAreaView>
        );
    }

    const ListItem = ({ item }) => {
        return (
            <TouchableOpacity
                onPress={() => navigation.navigate("TestScreen2", item.id)}
            // onPress={() => navigation.navigate("ProfileScreen")}
            >
                <View style={styles.item}>
                    <Image
                        source={{
                            uri: item.uri,
                        }}
                        style={styles.itemPhoto}
                        resizeMode="cover"
                    />
                    {/* <>{console.log(item)}</> */}
                    <Text style={styles.itemText}>{item.title}</Text>
                    {/* <Text style={styles.itemText}>{item.id}</Text> */}
                </View>
            </TouchableOpacity>
        );
    };

    // const renderItem = ({ item, index, section }) => <Text key={index}>{item.title}</Text>

    // const renderItem = ({ item, index }) => (
    //     <TouchableOpacity key={index}
    //         onPress={() => navigation.navigate("TestScreen2", item.id)}
    //     // onPress={() => navigation.navigate("ProfileScreen")}
    //     >
    //         <ListItem item={item} />
    //     </TouchableOpacity>
    // )

    return (
        // <SafeAreaView>
        <View>
            <SectionList
                // horizontal
                renderSectionHeader={({ section, index }) => (
                    <>
                        <Text style={styles.sectionHeader}>{section.title}</Text>
                        <FlatList
                            horizontal
                            data={section.data}
                            // renderItem={renderItem}
                            renderItem={({ item }) => <ListItem item={item} />}
                            showsHorizontalScrollIndicator={false}
                            keyExtractor={(item, index) => item + index}
                        />
                    </>
                )}
                renderItem={() => {
                    return null;
                    // return <ListItem item={item} />;
                }}
                // renderItem={renderItem}
                // renderSectionHeader={({ section: { title } }) => (
                //     <Text style={{ fontWeight: 'bold' }}>{title}</Text>
                // )}
                // renderSectionHeader={({ section: { title } }) => (
                //     <Text style={{ fontWeight: 'bold' }}>{title}</Text>
                // )}
                sections={categories}
                keyExtractor={(item, index) => item + index}
            />
            <Button title='Show useEffect' onPress={() => console.log("Button", categories)} />
        </View>
        // </SafeAreaView>
    )
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: "#121212",
    },
    sectionHeader: {
        fontWeight: "800",
        fontSize: 18,
        // color: "#f4f4f4",
        marginTop: 20,
        marginLeft: 10,
        marginBottom: 5,
    },
    item: {
        margin: 10,
    },
    itemPhoto: {
        width: 200,
        height: 200,
    },
    itemText: {
        // color: "rgba(255, 255, 255, 0.5)",
        marginTop: 5,
    },
});

export default CategorySectionList