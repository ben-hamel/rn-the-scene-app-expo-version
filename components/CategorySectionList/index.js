import CategorySectionList from "./CategorySectionList";

export default CategorySectionList;