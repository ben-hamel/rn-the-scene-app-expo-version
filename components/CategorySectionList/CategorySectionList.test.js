import React from "react";
import { render } from "@testing-library/react-native";
import CategorySectionList from "./CategorySectionList";

jest.mock("@react-navigation/native", () => ({
  ...jest.requireActual("@react-navigation/native"),
  useNavigation: () => ({
    navigate: jest.fn(),
  }),
}));

describe("CategorySectionList", () => {
  const categories = [
    {
      title: "Category 1",
      data: [
        { title: "Item 1", uri: "https://example.com/item1.jpg" },
        { title: "Item 2", uri: "https://example.com/item2.jpg" },
      ],
    },
    {
      title: "Category 2",
      data: [
        { title: "Item 3", uri: "https://example.com/item3.jpg" },
        { title: "Item 4", uri: "https://example.com/item4.jpg" },
      ],
    },
  ];

  it("renders loading state correctly", () => {
    const { getByText } = render(
      <CategorySectionList categories={[]} loading={true} />
    );
    expect(getByText("Loading")).toBeDefined();
  });

  it("renders category sections correctly", () => {
    const { getByText, getByTestId } = render(
      <CategorySectionList categories={categories} loading={false} />
    );

    categories.forEach((category) => {
      // Check if section headers are rendered
      expect(getByText(category.title)).toBeDefined();

      category.data.forEach((item) => {
        // Check if item titles are rendered
        expect(getByText(item.title)).toBeDefined();

        // Check if item images are rendered
        expect(getByTestId(`item-image-${item.title}`)).toBeDefined();
      });
    });
  });

  it("matches snapshot", () => {
    const { toJSON } = render(
      <CategorySectionList categories={categories} loading={false} />
    );

    // Compare the rendered component with a snapshot
    expect(toJSON()).toMatchSnapshot();
  });
});
