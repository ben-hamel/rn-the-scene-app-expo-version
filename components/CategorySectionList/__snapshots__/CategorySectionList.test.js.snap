// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategorySectionList matches snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={
      [
        {
          "data": [
            {
              "title": "Item 1",
              "uri": "https://example.com/item1.jpg",
            },
            {
              "title": "Item 2",
              "uri": "https://example.com/item2.jpg",
            },
          ],
          "title": "Category 1",
        },
        {
          "data": [
            {
              "title": "Item 3",
              "uri": "https://example.com/item3.jpg",
            },
            {
              "title": "Item 4",
              "uri": "https://example.com/item4.jpg",
            },
          ],
          "title": "Category 2",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={
      [
        0,
        4,
      ]
    }
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text
          style={
            [
              {
                "fontSize": 18,
                "fontWeight": "800",
                "marginBottom": 5,
                "marginLeft": 10,
                "marginTop": 20,
              },
              {
                "color": "rgb(28, 28, 30)",
              },
              {
                "textTransform": "capitalize",
              },
            ]
          }
        >
          Category 1
        </Text>
        <RCTScrollView
          data={
            [
              {
                "title": "Item 1",
                "uri": "https://example.com/item1.jpg",
              },
              {
                "title": "Item 2",
                "uri": "https://example.com/item2.jpg",
              },
            ]
          }
          getItem={[Function]}
          getItemCount={[Function]}
          horizontal={true}
          keyExtractor={[Function]}
          onContentSizeChange={[Function]}
          onLayout={[Function]}
          onMomentumScrollBegin={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEventThrottle={0.0001}
          showsHorizontalScrollIndicator={false}
          stickyHeaderIndices={[]}
          viewabilityConfigCallbackPairs={[]}
        >
          <View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={
                [
                  {
                    "flexDirection": "row",
                  },
                  null,
                ]
              }
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
              >
                <View
                  style={
                    {
                      "margin": 10,
                    }
                  }
                >
                  <Image
                    accessibilityRole="image"
                    alt="Item 1"
                    borderRadius={10}
                    resizeMode="cover"
                    source={
                      {
                        "uri": "https://example.com/item1.jpg",
                      }
                    }
                    style={
                      {
                        "height": 200,
                        "width": 200,
                      }
                    }
                    testID="item-image-Item 1"
                  />
                  <Text
                    style={
                      [
                        {
                          "color": "rgb(28, 28, 30)",
                          "marginTop": 5,
                        },
                        {
                          "textTransform": "capitalize",
                        },
                      ]
                    }
                  >
                    Item 1
                  </Text>
                </View>
              </View>
            </View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={
                [
                  {
                    "flexDirection": "row",
                  },
                  null,
                ]
              }
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
              >
                <View
                  style={
                    {
                      "margin": 10,
                    }
                  }
                >
                  <Image
                    accessibilityRole="image"
                    alt="Item 2"
                    borderRadius={10}
                    resizeMode="cover"
                    source={
                      {
                        "uri": "https://example.com/item2.jpg",
                      }
                    }
                    style={
                      {
                        "height": 200,
                        "width": 200,
                      }
                    }
                    testID="item-image-Item 2"
                  />
                  <Text
                    style={
                      [
                        {
                          "color": "rgb(28, 28, 30)",
                          "marginTop": 5,
                        },
                        {
                          "textTransform": "capitalize",
                        },
                      ]
                    }
                  >
                    Item 2
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      />
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      />
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      />
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text
          style={
            [
              {
                "fontSize": 18,
                "fontWeight": "800",
                "marginBottom": 5,
                "marginLeft": 10,
                "marginTop": 20,
              },
              {
                "color": "rgb(28, 28, 30)",
              },
              {
                "textTransform": "capitalize",
              },
            ]
          }
        >
          Category 2
        </Text>
        <RCTScrollView
          data={
            [
              {
                "title": "Item 3",
                "uri": "https://example.com/item3.jpg",
              },
              {
                "title": "Item 4",
                "uri": "https://example.com/item4.jpg",
              },
            ]
          }
          getItem={[Function]}
          getItemCount={[Function]}
          horizontal={true}
          keyExtractor={[Function]}
          onContentSizeChange={[Function]}
          onLayout={[Function]}
          onMomentumScrollBegin={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEventThrottle={0.0001}
          showsHorizontalScrollIndicator={false}
          stickyHeaderIndices={[]}
          viewabilityConfigCallbackPairs={[]}
        >
          <View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={
                [
                  {
                    "flexDirection": "row",
                  },
                  null,
                ]
              }
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
              >
                <View
                  style={
                    {
                      "margin": 10,
                    }
                  }
                >
                  <Image
                    accessibilityRole="image"
                    alt="Item 3"
                    borderRadius={10}
                    resizeMode="cover"
                    source={
                      {
                        "uri": "https://example.com/item3.jpg",
                      }
                    }
                    style={
                      {
                        "height": 200,
                        "width": 200,
                      }
                    }
                    testID="item-image-Item 3"
                  />
                  <Text
                    style={
                      [
                        {
                          "color": "rgb(28, 28, 30)",
                          "marginTop": 5,
                        },
                        {
                          "textTransform": "capitalize",
                        },
                      ]
                    }
                  >
                    Item 3
                  </Text>
                </View>
              </View>
            </View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={
                [
                  {
                    "flexDirection": "row",
                  },
                  null,
                ]
              }
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
              >
                <View
                  style={
                    {
                      "margin": 10,
                    }
                  }
                >
                  <Image
                    accessibilityRole="image"
                    alt="Item 4"
                    borderRadius={10}
                    resizeMode="cover"
                    source={
                      {
                        "uri": "https://example.com/item4.jpg",
                      }
                    }
                    style={
                      {
                        "height": 200,
                        "width": 200,
                      }
                    }
                    testID="item-image-Item 4"
                  />
                  <Text
                    style={
                      [
                        {
                          "color": "rgb(28, 28, 30)",
                          "marginTop": 5,
                        },
                        {
                          "textTransform": "capitalize",
                        },
                      ]
                    }
                  >
                    Item 4
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      />
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      />
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      />
    </View>
  </RCTScrollView>
</View>
`;
